language/vm/src/proptest_types/types.rs

```diff
@@ -0,0 +1,229 @@
+    },
+    proptest_types::signature::{KindGen, SignatureTokenGen},
+};
+use proptest::{
+    collection::{vec, SizeRange},
+    option,
+    prelude::*,
+    sample::Index as PropIndex,
+    std_facade::hash_set::HashSet,
+};
+use std::collections::BTreeSet;
+
+#[derive(Debug)]
+struct TypeSignatureIndex(u16);
+
+#[derive(Debug)]
+pub struct StDefnMaterializeState {
+    pub identifiers_len: usize,
+    pub struct_handles: Vec<StructHandle>,
+    pub new_handles: BTreeSet<(ModuleHandleIndex, IdentifierIndex)>,
+}
+
+impl StDefnMaterializeState {
+    pub fn new(identifiers_len: usize, struct_handles: Vec<StructHandle>) -> Self {
+        Self {
+            identifiers_len,
+            struct_handles,
+            new_handles: BTreeSet::new(),
+        }
+    }
+
+    fn add_struct_handle(&mut self, handle: StructHandle) -> Option<StructHandleIndex> {
+        if self.new_handles.insert((handle.module, handle.name)) {
+            self.struct_handles.push(handle);
+            Some(StructHandleIndex((self.struct_handles.len() - 1) as u16))
+        } else {
+            None
+        }
+    }
+
+    fn contains_nominal_resource(&self, signature: &SignatureToken) -> bool {
+        use SignatureToken::*;
+
+        match signature {
+            Signer => true,
+            Struct(struct_handle_index) => {
+                self.struct_handles[struct_handle_index.0 as usize].is_nominal_resource
+            }
+            StructInstantiation(struct_handle_index, type_args) => {
+                self.struct_handles[struct_handle_index.0 as usize].is_nominal_resource
+                    || type_args.iter().any(|t| self.contains_nominal_resource(t))
+            }
+            Vector(targ) => self.contains_nominal_resource(targ),
+            Reference(token) | MutableReference(token) => self.contains_nominal_resource(token),
+            Bool | U8 | U64 | U128 | Address | TypeParameter(_) => false,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct StructHandleGen {
+    module_idx: PropIndex,
+    name_idx: PropIndex,
+    is_nominal_resource: bool,
+    type_parameters: Vec<KindGen>,
+}
+
+impl StructHandleGen {
+    pub fn strategy(kind_count: impl Into<SizeRange>) -> impl Strategy<Value = Self> {
+        (
+            any::<PropIndex>(),
+            any::<PropIndex>(),
+            any::<bool>(),
+            vec(KindGen::strategy(), kind_count),
+        )
+            .prop_map(
+                |(module_idx, name_idx, is_nominal_resource, type_parameters)| Self {
+                    module_idx,
+                    name_idx,
+                    is_nominal_resource,
+                    type_parameters,
+                },
+            )
+    }
+
+    pub fn materialize(self, module_len: usize, identifiers_len: usize) -> StructHandle {
+        let mut idx = self.module_idx.index(module_len) as TableIndex;
+        if idx == 0 {
+            idx += 1;
+        }
```

Comment:
not sure I understand `min`, we are just skipping 0 here. Basically we don't want to create any `StructHandle` with `0` because the `StructDefinition` will make them. 
We are making all the type imports here

Reply:
Oops, should be max :p 

Unless I'm missing something, your code is equivalent to 
```
let idx = {
    let x = self.module_idx.index(module_len) as TableIndex;
    if x == 0 {
        1
    } else {
        x 
    }
};
```

Which is the same as `max(..., 1)`